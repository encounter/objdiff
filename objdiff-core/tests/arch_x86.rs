use objdiff_core::{diff, diff::display::SymbolFilter, obj};

mod common;

#[test]
#[cfg(feature = "x86")]
fn read_x86() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/staticdebug.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx = obj.symbols.iter().position(|s| s.name == "?PrintThing@@YAXXZ").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

#[test]
#[cfg(feature = "x86")]
fn read_x86_combine_sections() {
    let diff_config = diff::DiffObjConfig {
        combine_data_sections: true,
        combine_text_sections: true,
        ..Default::default()
    };
    let obj = obj::read::parse(include_object!("data/x86/rtest.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj.sections);
}

#[test]
#[cfg(feature = "x86")]
fn read_x86_64() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86_64/vs2022.o"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx =
        obj.symbols.iter().position(|s| s.name == "?Dot@Vector@@QEAAMPEAU1@@Z").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

#[test]
#[cfg(feature = "x86")]
fn display_section_ordering() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/basenode.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    let obj_diff =
        diff::diff_objs(Some(&obj), None, None, &diff_config, &diff::MappingConfig::default())
            .unwrap()
            .left
            .unwrap();
    let section_display =
        diff::display::display_sections(&obj, &obj_diff, SymbolFilter::None, false, false, false);
    insta::assert_debug_snapshot!(section_display);
}

#[test]
#[cfg(feature = "x86")]
fn read_x86_jumptable() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/jumptable.o"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx = obj.symbols.iter().position(|s| s.name == "?test@@YAHH@Z").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

// Inferred size of functions should ignore symbols with specific prefixes
#[test]
#[cfg(feature = "x86")]
fn read_x86_local_labels() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/local_labels.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
}
