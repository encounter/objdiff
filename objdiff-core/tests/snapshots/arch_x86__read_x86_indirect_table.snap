---
source: objdiff-core/tests/arch_x86.rs
expression: obj
---
Object {
    arch: ArchX86 {
        arch: X86,
        endianness: Little,
    },
    endianness: Little,
    symbols: [
        Symbol {
            name: "C:\\Dev\\Projects\\calineva-legacy\\src\\game\\control\\bridge.cpp",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "@comp.id",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Object,
            section: None,
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.drectve]",
            demangled_name: None,
            address: 0,
            size: 44,
            kind: Section,
            section: Some(
                0,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.debug$S]",
            demangled_name: None,
            address: 0,
            size: 271,
            kind: Section,
            section: Some(
                1,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.text]",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Section,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.debug$S]",
            demangled_name: None,
            address: 0,
            size: 620,
            kind: Section,
            section: Some(
                3,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "c:\\dev\\projects\\calineva-legacy\\src\\game\\control\\bridge.cpp",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "?process@@YAHHHH@Z",
            demangled_name: Some(
                "int __cdecl process(int, int, int)",
            ),
            address: 0,
            size: 1329,
            kind: Function,
            section: Some(
                2,
            ),
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L296",
            demangled_name: None,
            address: 418,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L312",
            demangled_name: None,
            address: 217,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L294",
            demangled_name: None,
            address: 406,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L293",
            demangled_name: None,
            address: 389,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L292",
            demangled_name: None,
            address: 369,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L291",
            demangled_name: None,
            address: 360,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L332",
            demangled_name: None,
            address: 1104,
            size: 0,
            kind: Object,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L327",
            demangled_name: None,
            address: 1128,
            size: 0,
            kind: Object,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L289",
            demangled_name: None,
            address: 311,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L288",
            demangled_name: None,
            address: 300,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L287",
            demangled_name: None,
            address: 291,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L286",
            demangled_name: None,
            address: 282,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L285",
            demangled_name: None,
            address: 273,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L284",
            demangled_name: None,
            address: 264,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L331",
            demangled_name: None,
            address: 824,
            size: 0,
            kind: Object,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L326",
            demangled_name: None,
            address: 852,
            size: 0,
            kind: Object,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L311",
            demangled_name: None,
            address: 205,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L310",
            demangled_name: None,
            address: 185,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L309",
            demangled_name: None,
            address: 175,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L308",
            demangled_name: None,
            address: 165,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L330",
            demangled_name: None,
            address: 652,
            size: 0,
            kind: Object,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L325",
            demangled_name: None,
            address: 672,
            size: 0,
            kind: Object,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L306",
            demangled_name: None,
            address: 102,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L305",
            demangled_name: None,
            address: 88,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L304",
            demangled_name: None,
            address: 79,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L302",
            demangled_name: None,
            address: 70,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L329",
            demangled_name: None,
            address: 424,
            size: 0,
            kind: Object,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$L324",
            demangled_name: None,
            address: 448,
            size: 0,
            kind: Object,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: ".bf",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: ".lf",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: ".ef",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.debug$F]",
            demangled_name: None,
            address: 0,
            size: 16,
            kind: Section,
            section: Some(
                4,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.debug$T]",
            demangled_name: None,
            address: 0,
            size: 104,
            kind: Section,
            section: Some(
                5,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
    ],
    sections: [
        Section {
            id: ".drectve-0",
            name: ".drectve",
            address: 0,
            size: 44,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                1,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".debug$S-0",
            name: ".debug$S",
            address: 0,
            size: 271,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                1,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".text-0",
            name: ".text",
            address: 0,
            size: 1344,
            kind: Code,
            data: SectionData(
                1344,
            ),
            flags: FlagSet(),
            align: Some(
                16,
            ),
            relocations: [
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 59,
                    target_symbol: 35,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 66,
                    target_symbol: 34,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 154,
                    target_symbol: 29,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 161,
                    target_symbol: 28,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 253,
                    target_symbol: 23,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 260,
                    target_symbol: 22,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 349,
                    target_symbol: 15,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 356,
                    target_symbol: 14,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 424,
                    target_symbol: 33,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 428,
                    target_symbol: 13,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 432,
                    target_symbol: 32,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 436,
                    target_symbol: 31,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 440,
                    target_symbol: 30,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 444,
                    target_symbol: 9,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 652,
                    target_symbol: 27,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 656,
                    target_symbol: 26,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 660,
                    target_symbol: 25,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 664,
                    target_symbol: 24,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 668,
                    target_symbol: 9,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 824,
                    target_symbol: 21,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 828,
                    target_symbol: 20,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 832,
                    target_symbol: 19,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 836,
                    target_symbol: 18,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 840,
                    target_symbol: 17,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 844,
                    target_symbol: 16,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 848,
                    target_symbol: 8,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 1104,
                    target_symbol: 13,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 1108,
                    target_symbol: 12,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 1112,
                    target_symbol: 11,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 1116,
                    target_symbol: 10,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 1120,
                    target_symbol: 33,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 1124,
                    target_symbol: 8,
                    addend: 0,
                },
            ],
            line_info: {
                0: 9,
                21: 91,
                28: 94,
                29: 55,
                70: 57,
                78: 94,
                79: 63,
                87: 94,
                88: 66,
                101: 94,
                102: 69,
                123: 94,
                124: 72,
                138: 94,
                139: 55,
                165: 75,
                174: 94,
                175: 78,
                184: 94,
                185: 81,
                204: 94,
                205: 84,
                216: 94,
                217: 87,
                222: 94,
                223: 12,
                264: 14,
                272: 94,
                273: 17,
                281: 94,
                282: 20,
                290: 94,
                291: 23,
                299: 94,
                300: 26,
                310: 94,
                311: 29,
                322: 94,
                323: 32,
                333: 94,
                334: 12,
                360: 35,
                368: 94,
                369: 38,
                388: 94,
                389: 41,
                405: 94,
                406: 44,
                417: 94,
                418: 50,
                421: 94,
            },
            virtual_address: None,
        },
        Section {
            id: ".debug$S-1",
            name: ".debug$S",
            address: 0,
            size: 620,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                1,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".debug$F-0",
            name: ".debug$F",
            address: 0,
            size: 16,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                1,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".debug$T-0",
            name: ".debug$T",
            address: 0,
            size: 104,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                1,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
    ],
    split_meta: None,
    path: None,
    timestamp: None,
    flow_analysis_results: {},
}
